vault/resource_userpass_user.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+)
+
+func userpassUserResource() *schema.Resource {
+	fields := map[string]*schema.Schema{
+		consts.FieldMount: {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			Description:  "Auth mount at which user will be configured.",
+			Default:      consts.MountTypeUserpass,
+			ValidateFunc: provider.ValidateNoLeadingTrailingSlashes,
+		},
+		consts.FieldUsername: {
+			Type:         schema.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			Description:  "Username part of userpass.",
+			ValidateFunc: provider.ValidateStringSlug,
+		},
+		consts.FieldPassword: {
+			Type:        schema.TypeString,
+			Required:    true,
+			Sensitive:   true,
+			ForceNew:    true,
+			Description: "Password part of userpass.",
+		},
+	}
+
+	addTokenFields(fields, &addTokenFieldsConfig{})
+
+	return &schema.Resource{
+		CreateContext: userpassUserCreate,
+		ReadContext:   ReadContextWrapper(userpassUserRead),
+		UpdateContext: userpassUserUpdate,
+		DeleteContext: userpassUserDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: fields,
+	}
+}
+
+var (
+	userpassMountFromPathRegex    = regexp.MustCompile("^auth/(.+)/users/.+$")
+	userpassUsernameFromPathRegex = regexp.MustCompile("^auth/.+/users/(.+)$")
+)
+
+func userPath(mount string, username string) string {
+	return fmt.Sprintf("auth/%s/users/%s", mount, username)
+}
+
+func mountFromPath(path string) (string, error) {
+	if !userpassMountFromPathRegex.MatchString(path) {
+		return "", fmt.Errorf("no mount found")
+	}
+	res := userpassMountFromPathRegex.FindStringSubmatch(path)
+	if len(res) != 2 {
+		return "", fmt.Errorf("unexpected number of matches (%d) for mount", len(res))
+	}
+	return res[1], nil
+}
+
+func usernameFromPath(path string) (string, error) {
+	if !userpassUsernameFromPathRegex.MatchString(path) {
+		return "", fmt.Errorf("no name found")
+	}
+	res := userpassUsernameFromPathRegex.FindStringSubmatch(path)
+	if len(res) != 2 {
+		return "", fmt.Errorf("unexpected number of matches (%d) for username", len(res))
+	}
+	return res[1], nil
+}
+
+func userpassUserCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	id := userPath(d.Get(consts.FieldMount).(string), d.Get(consts.FieldUsername).(string))
+	d.SetId(id)
+	d.MarkNewResource()
```

Comment:
We can remove this from here since this would be the default behavior for this method anyway.
```suggestion

```